### PR TITLE
[ImageClassifier] Add an option to convert the input and output to fp16

### PR DESCRIPTION
**Pull request by qcolombet**
[ImageClassifier] Add a convert-inout-to-fp16 option
    
    *Description*
    This teaches the image-classifier how to convert the input and output
    of a network to fp16.
    
    *Testing*
    Ran resnet50 with -convert-inout-to-fp16

### CODE REVIEW

**Comment by qcolombet** [comment]
@opti-mix This patch should do what you want, modulo, you need you're commit with the optimization that removes the conversion between the same types.

This is WIP because I need to add the tests for the placeholder conversion. But the code itself should be done.

**Comment by qcolombet** [comment]
Added test case.

File: lib/Converter/FunctionConverter.cpp [hunk 1]
```diff
@@ -59,6 +62,99 @@ Node &FunctionConverter::morphNode(Node &node) { return node; }
...
+           "Conversion does not preserve shape");
+    TypeRef origTy = val.getType();
+    // Fake the morphing of the node so that the creation
+    // of the conversion works properly.
+    val.setType(targetTy);
+    // Store the users in a temporary object because setOperand
+    // will invalidate the iterator.
+    llvm::SmallVector<NodeUse, 4> users(val.getUsers().begin(),
+                                        val.getUsers().end());
+    // We cannot use replaceAllUsesWith here because:
```
Comment (beicy) [inline lib/Converter/FunctionConverter.cpp]: I noticed this "4 users" before but forget to ask -- Just wondering how this "4" comes from :) Thanks!

File: lib/Converter/FunctionConverter.cpp [hunk 2]
```diff
@@ -59,6 +62,99 @@ Node &FunctionConverter::morphNode(Node &node) { return node; }
...
+    //   "conversion".
+    // 3. Node may be a module level value and we need one conversion per
+    //    function.
+    for (auto use : users) {
+      Node *user = use.getUser();
+      Function *parent = user->getParent();
+      assert(parent && "User not in a function?!");
+
+      SaveNode *saveNode = llvm::dyn_cast<SaveNode>(user);
+      if (saveNode) {
```
[the remark this answers is not in the document]
Comment (qcolombet) [reply lib/Converter/FunctionConverter.cpp]: Good catch, that's a relic of some specialization I removed from the code.

File: tools/loader/ImageClassifier.cpp [hunk 3]
```diff
@@ -129,6 +130,12 @@ llvm::cl::opt<std::string> modelInputName(
     llvm::cl::desc("The name of the variable for the model's input image."),
     llvm::cl::value_desc("string_name"), llvm::cl::Required,
     llvm::cl::cat(imageLoaderCat));
+
+llvm::cl::opt<bool> convertInputAndOutToFp16(
```
[the remark this answers is not in the document]
Comment (qcolombet) [reply tools/loader/ImageClassifier.cpp]: Done.

File: lib/Converter/FunctionConverter.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (qcolombet) [reply lib/Converter/FunctionConverter.cpp]: Done

File: tools/loader/ImageClassifier.cpp [hunk 4]
```diff
@@ -299,6 +315,9 @@ int main(int argc, char **argv) {
 
     // Perform the inference execution.
     loader.runInference(ctx);
+    if (convertInAndOutToFp16) {
+      SMVarT->convertToType(ElemKind::FloatTy);
```
Comment (jfix71) [inline tools/loader/ImageClassifier.cpp]: Shouldn't this be `ElemKind::Float16Ty`?

Comment (opti-mix) [reply tools/loader/ImageClassifier.cpp]: @jfix71 If I understand the code correctly, then everything is fine. We convert the payload of the tensor containing final results from fp16 into fp32 and then use the usual fp32-based function afterwards to print those results. May be a comment before the `if` explaining what's going on would be useful.

Comment (qcolombet) [reply tools/loader/ImageClassifier.cpp]: That is correct.
I've added a comment to explain that. Let me know what you think.

Comment (jfix71) [reply tools/loader/ImageClassifier.cpp]: Looks good, thanks!

**Pull request merged**